compiler/qsc_eval/src/val.rs

```diff
@@ -23,11 +23,65 @@ pub enum Value {
     Pauli(Pauli),
     Qubit(Qubit),
     Range(Option<i64>, i64, Option<i64>),
-    Result(bool),
+    Result(Result),
     String(Rc<str>),
     Tuple(Rc<[Value]>),
 }
 
+#[derive(Clone, Copy, Debug)]
+pub enum Result {
+    Val(bool),
+    Id(usize),
+}
+
+impl From<Result> for bool {
+    fn from(val: Result) -> Self {
+        match val {
+            Result::Val(v) => v,
+            Result::Id(_) => panic!("cannot convert Result::Id to bool"),
+        }
+    }
+}
+
+impl From<&Result> for bool {
+    fn from(val: &Result) -> Self {
+        match val {
+            Result::Val(v) => *v,
+            Result::Id(_) => panic!("cannot convert Result::Id to bool"),
+        }
+    }
+}
+
+impl From<bool> for Result {
+    fn from(val: bool) -> Self {
+        Self::Val(val)
+    }
+}
+
+impl From<usize> for Result {
+    fn from(val: usize) -> Self {
+        Self::Id(val)
+    }
+}
+
+impl From<Result> for usize {
+    fn from(val: Result) -> Self {
+        match val {
+            Result::Val(..) => panic!("cannot convert Result::Val to usize"),
```

Comment:
> Note: This trait must not fail. The [From](https://doc.rust-lang.org/std/convert/trait.From.html) trait is intended for perfect conversions. If the conversion can fail or is not perfect, use [TryFrom](https://doc.rust-lang.org/std/convert/trait.TryFrom.html).

Reply:
(And yeah, I know NodeId is also doing sketchy things with From and PartialEq. That bothers me too and I think the default node ID stuff should probably be refactored.)

Reply:
I understand. But that's not what the From trait is for.

Reply:
Fair enough. It moves the panic message from one place in the implementation up to each of the callers and forces the use of `expect` there, which I don't prefer. But I made the change anyway.

Reply:
You don't need to move the panic message, remember the [`unwrap_foo` pattern](https://github.com/microsoft/qsharp/blob/d0f77e0c94c44045bf916fa753d93019552d576c/compiler/qsc_eval/src/val.rs#L234)?

Reply:
And here I thought you were recommending `TryFrom` again. Yes, I can use the pattern we used elsewhere with values.

Reply:
No, sorry, my first comment was just quoting the docs for From. I wasn't suggesting that you follow its suggestion literally. My point was only that From shouldn't be used when the conversion is fallible.